### PR TITLE
[db_migrator]Do DB migration for buffer pool size change on Mellanox platform with new SDK version

### PR DESCRIPTION
**Pull request by keboliu**
**- What I did**
Do DB migration for buffer pool size change with the new SDK version, this is specifically for the Mellanox platform.

**- How I did it**
Advance the DB version to a new version 1_0_3, and add Mellanox specific DB migration functions.
Inside the function to migrate the buffer pool size to the latest values.

**- How to verify it**
test sonic to sonic upgrade from earlier 201911 version to version with the latest SDK.
test sonic to sonic upgrade from 201811 to 201911 version with the latest SDK.
test upgrade with different topology.

**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**

### CODE REVIEW

File: scripts/db_migrator.py [hunk 1]
```diff
@@ -127,13 +211,22 @@ def version_1_0_1(self):
 
         self.migrate_interface_table()
         self.set_version('version_1_0_2')
-        return None
+        return 'version_1_0_2'
```
Comment (nazariig) [inline scripts/db_migrator.py]: @keboliu why the None is missing here?

File: scripts/db_migrator.py [hunk 2]
```diff
@@ -96,6 +100,86 @@ def migrate_interface_table(self):
...
+
+        if pool_size_in_db_list == spc1_t0_default_value or pool_size_in_db_list == spc1_t1_default_value \
+            or pool_size_in_db_list == spc2_t0_default_value or pool_size_in_db_list == spc2_t1_default_value:
+            migration_needed = True
+
+        if migration_needed:
+            # Generate buffer configuration from latest template
+            tmp_json_file = "/tmp/mlnx_buffers_all.json"
+            device_buffer_template_path = os.path.join("/usr/share/sonic/device/", platform, hwsku, 'buffers.json.j2')
+            command = "sonic-cfggen -d -t " + device_buffer_template_path + " >" + tmp_json_file
```
Comment (nazariig) [inline scripts/db_migrator.py]: @keboliu consider using format

File: scripts/db_migrator.py [hunk 3]
```diff
@@ -96,6 +100,86 @@ def migrate_interface_table(self):
...
+            proc = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE)
```
Comment (nazariig) [inline scripts/db_migrator.py]: @keboliu consider using check_call

File: scripts/db_migrator.py [hunk 4]
```diff
@@ -96,6 +100,86 @@ def migrate_interface_table(self):
...
+            # Extract the new buffer pool size configuration from above tmp file and update DB
+            try:
+                with open(tmp_json_file) as tmp_buffer_file:
+                    buffer_json = json.load(tmp_buffer_file)
+                    for pool in buffer_pools:
+                        self.configDB.set_entry('BUFFER_POOL', pool, buffer_json['BUFFER_POOL'][pool])
+                    log_info("Successfully migrate buffer pool size to the latest.")
+            except IOError:
+                log_error("failed to open tmp mlnx buffer json configuration file, skip migration")
+                return
```
Comment (nazariig) [inline scripts/db_migrator.py]: @keboliu why do we need return here?

File: scripts/db_migrator.py (hunk 1 above)
Comment (keboliu) [reply scripts/db_migrator.py]: after migrating to V 1_0_1, we need to drive db_migrator to proceed to migrate from 1_0_2 to 1_0_3,  check the logic in function "migrate()", a None return value will stop it from migrating to next version.

Comment (keboliu) [reply scripts/db_migrator.py]: after migrating to V 1_0_2, we need to drive db_migrator to proceed to migrate from 1_0_2 to 1_0_3,  check the logic in function "migrate()", a None return value will stop it from migrating to next version.

File: scripts/db_migrator.py (hunk 4 above)
Comment (keboliu) [reply scripts/db_migrator.py]: make sense, can be removed.

File: scripts/db_migrator.py (hunk 3 above)
Comment (keboliu) [reply scripts/db_migrator.py]: make sense, since we don't need to output here.

File: scripts/db_migrator.py (hunk 1 above)
Comment (keboliu) [reply scripts/db_migrator.py]: after migrating to V 1_0_1, we need to drive db_migrator to proceed to migrate from 1_0_2 to 1_0_3,  check the logic in function "migrate()", a None return value will stop it from migrating to next version.

File: scripts/db_migrator.py (hunk 4 above)
Comment (keboliu) [reply scripts/db_migrator.py]: changed

**Comment by keboliu** [comment]
> @keboliu, have you tested these changes with the warm-reboot?
> Just want to make sure that we will not have any problems with that since config is going to be changed in new image comparing to old one

If my understanding is correct, this whole DB migration process will be skipped during the warm reboot case,  this fix is only for sonic to sonic upgrade flow. I think we have another thread discussing how to handle this change in the warm reboot case?

File: scripts/db_migrator.py [hunk 5]
```diff
@@ -218,8 +212,13 @@ def version_1_0_2(self):
         Version 1_0_2.
         """
         log_info('Handling version_1_0_2')
-        self.mlnx_migrate_buffer_pool_size()
-        self.set_version('version_1_0_3')
+        # Check ASIC type, if Mellanox platform then need DB migration
+        version_info = sonic_device_util.get_sonic_version_info()
+        if version_info['asic_type'] == "mellanox":
+            if self.mlnx_migrate_buffer_pool_size():
```
Comment (nazariig) [inline scripts/db_migrator.py]: @keboliu what will happen if self.mlnx_migrate_buffer_pool_size() return False? Looks like target version won't be set.

Comment (keboliu) [reply scripts/db_migrator.py]: if it returns fail means something wrong during the migration, in the current implementation it probably due to lack of "config_db.json", failed to generate new buffer config or something like this. What I try to say is that all the case return fail is that the DB migration should be success by failed, so in these cases, we don't change the DB version.

**Pull request closed without merging**